shiny/express/_run.py

```diff
@@ -0,0 +1,243 @@
+from .display_decorator._node_transformers import (
+    DisplayFuncsTransformer,
+    display_decorator_func_name,
+)
+
+__all__ = (
+    "wrap_express_app",
+    "is_express_app",
+)
+
+_DEFAULT_PAGE_FUNCTION = ui.page_fluid
+
+
+def wrap_express_app(file: Path | None = None) -> App:
+    """Wrap a Shiny express-mode app into a Shiny `App` object.
+
+    Parameters
+    ----------
+    file
+        The path to the file containing the Shiny express application. If `None`, the
+        `SHINY_EXPRESS_APP_FILE` environment variable is used.
+
+    Returns
+    -------
+    :
+        A `shiny.App` object.
+    """
+    if file is None:
+        app_file = os.getenv("SHINY_EXPRESS_APP_FILE")
+        if app_file is None:
+            raise ValueError(
+                "No app file was specified and the SHINY_EXPRESS_APP_FILE environment variable "
+                "is not set."
+            )
+        file = Path(os.getcwd()) / app_file
+
+    # TODO: title and lang
+    app_ui = ui.page_output("__page__")
+
+    def express_server(input: Inputs, output: Outputs, session: Session):
+        try:
+            dyn_ui = run_express(file)
+
+            @render.ui
+            def __page__():
+                return dyn_ui
+
+        except Exception:
+            import traceback
+
+            traceback.print_exception(*sys.exc_info())
+            raise
+
+    app = App(app_ui, express_server)
+
+    return app
+
+
+def is_express_app(app: str, app_dir: str | None) -> bool:
+    """Detect whether an app file is a Shiny express app
+
+    Parameters
+    ----------
+    app
+        App filename, like "app.py". It may be a relative path or absolute path.
+    app_dir
+        Directory containing the app file. If this is `None`, then `app` must be an
+        absolute path.
+
+    Returns
+    -------
+    :
+        `True` if it is a Shiny express app, `False` otherwise.
+    """
+    if not app.lower().endswith(".py"):
+        return False
+
+    if app_dir is not None:
+        app_path = Path(app_dir) / app
+    else:
+        app_path = Path(app)
+
+    if not app_path.exists():
+        return False
+
+    # Read the file, parse it, and look for any imports of shiny.express.
+    with open(app_path) as f:
+        content = f.read()
+    tree = ast.parse(content, app_path)
+    detector = DetectShinyExpressVisitor()
+    detector.visit(tree)
+
+    return detector.found_shiny_express_import
+
+
+def run_express(file: Path) -> Tag | TagList:
+    with open(file) as f:
+        content = f.read()
+
+    tree = ast.parse(content, file)
+    tree = DisplayFuncsTransformer().visit(tree)
+    tree = ast.fix_missing_locations(tree)
+
+    ui_result: Tag | TagList = TagList()
+
+    def set_result(x: object):
+        nonlocal ui_result
+        ui_result = cast(Tag, x)
+
+    sys.displayhook = set_result
+
+    reset_top_level_recall_context_manager()
+    get_top_level_recall_context_manager().__enter__()
+
+    file_path = str(file.resolve())
```

Comment:
Maybe require this to be absolute rather than resolving it here? Or not, I don't know.